crates/solana-event-forwarder/src/component.rs

```diff
@@ -53,38 +82,135 @@ impl SolanaEventForwarder {
         while let Some(message) = self.solana_listener_client.log_receiver.next().await {
             let gateway_program_stack =
                 build_program_event_stack(&match_context, &message.logs, parse_gateway_logs);
+            let gas_events_program_stack =
+                build_program_event_stack(&match_context, &message.logs, parse_gas_service_log);
+            // todo -- total cost is not representative
             let total_cost = message.cost_in_lamports;
 
             // Collect all successful events into a vector
-            let events_vec = gateway_program_stack
+            let gateway_events_vec = keep_successful_events(gateway_program_stack)
                 .into_iter()
-                .filter_map(|x| {
-                    if let ProgramInvocationState::Succeeded(events) = x {
-                        Some(events)
-                    } else {
-                        None
-                    }
-                })
-                .flatten()
-                .collect::<Vec<_>>();
+                .map(|(idx, event)| (idx, GatewayOrGasEvent::GatewayEvent(event)));
+            let gas_events = keep_successful_events(gas_events_program_stack)
+                .into_iter()
+                .map(|(idx, event)| (idx, GatewayOrGasEvent::GasEveent(event)));
+            let all_events = gateway_events_vec
+                .chain(gas_events)
+                .sorted_by(|event_a, event_b| event_a.0.cmp(&event_b.0));
+
+            // we have to associate the gas events with the gateway events because for
+            // NativeGasPaidForContractCallEvent event we need to attach the message_id that
+            // corresponds with the given CallContract event.
+            let combined_events = all_events
+                .fold(
+                    // (accumulated vector, pending NativeGasPaidForContractCallEvent)
+                    (vec![], Vec::<NativeGasPaidForContractCallEvent>::new()),
```

Comment:
using a Vec instead of a hashmap because iteration idxs is computationally cheaper than computing hashes in a hashmap. And we expect the data size to be small (usually 1-2 entries max)